### PR TITLE
Introducing flow with 2 levels of VLAN tagging on endpoints

### LINKED ISSUE #2949: no validation for the 'vlan_id' field while creating a flow via APIv2
**Step to reproduce**

- try to create a flow with negative vlan number(e.g. `-1`) via APIv2;

**Actual result**
Flow is created with negative vlan_id(` -1 `).
Then as result we can't GET flow(s) and delete the created flow.
500 Internal Server Error is returned to mentioned requests.

**Expected result**
Flow shouldn't be created.

**Workaround**
update the flow with valid `vlan-id`

### PR DESCRIPTION
**Pull request by surabujin**
This change adds the ability to use 2 levels of VLAN tags on flow endpoints.  I.e. you can define 2 (outer and inner) VLAN tags for each flow endpoint.

Flow's endpoints (source and dest) encoding are independent, so it can be defined as "default port" (no VLAN tags), as VLAN tagged and as QinQ or double VLAN tagged for each end of the flow.

OF spec allow to the switch to reject flows that manipulate with possible missing ethernet packet's field. If the OF flow's action list contains pop VLAN tag(for example) action it must have VLAN tag match in it's match field (see section 7.2.6 Action Structures OF specification v1.5.1 for more details). As a result we can't do the flow's endpoint encapsulation on egress switch as before, we need to unwrap ingress endpoint encapsulation on egress switch and apply/wrap egress encapsulation on egress switch.

So ingress switch will drop all(except last in case of transit-VLAN encapsulation) VLAN tags from ethernet packet, wrap it into transit encapsulation and on egress switch transit encapsulation will be unwrapped and corresponding egress endpoint VLAN tags are added.

---
This change is [Reviewable](https://reviewable.io/reviews/telstra/open-kilda/2962)

Fixes #2949

### CODE REVIEW

File: services/wfm/src/main/java/org/openkilda/wfm/share/logger/FlowOperationsDashboardLogger.java [hunk 1]
```diff
@@ -202,15 +207,18 @@ public void onFlowUpdate(Flow flow) {
     /**
      * Log a flow-update event.
      */
-    public void onFlowUpdate(String flowId, SwitchId srcSwitch, int srcPort, int srcVlan,
-                             SwitchId destSwitch, int destPort, int destVlan, String diverseFlowId, long bandwidth) {
+    public void onFlowUpdate(String flowId, RequestedFlow flow) {
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/share/logger/FlowOperationsDashboardLogger.java]: What is the reason to bring H&S specific class (RequestedFlow) into shared FlowOperationsDashboardLogger? After project restructuring, the logger can be a part of a shared module. But this dependency prevents this.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheBolt.java [hunk 2]
```diff
@@ -102,10 +108,11 @@ private void initFlowCache(FlowRepository flowRepository) {
                             log.warn("Flow {} has no meter ID", path.getFlow().getFlowId());
                         }
                     });
-            logger.debug("cookieToFlow cache: {}, switchAndMeterToFlow cache: {}", cookieToFlow, switchAndMeterToFlow);
-            logger.info("Stats Cache: Initialized");
+            log.debug(
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheBolt.java]: Again, shouldn't avoid logging the whole cache?

File: services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/RequestedFlowMapper.java [hunk 3]
```diff
@@ -31,18 +34,19 @@
...
-    @Mapping(source = "sourceVlan", target = "srcVlan")
-    @Mapping(source = "destinationSwitch", target = "destSwitch")
-    @Mapping(source = "destinationPort", target = "destPort")
-    @Mapping(source = "destinationVlan", target = "destVlan")
-    @Mapping(target = "flowEncapsulationType",
-            expression = "java(java.util.Optional.ofNullable(request.getEncapsulationType())"
-                    + ".map(v -> org.openkilda.model.FlowEncapsulationType.valueOf(v.toUpperCase()))"
-                    + ".orElse(null))")
-    public abstract RequestedFlow toRequestedFlow(FlowRequest request);
+    public RequestedFlow unpackRequest(FlowRequest request) {
```
Comment (sergii-iakovenko) [inline services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/RequestedFlowMapper.java]: Why "unpack"? I don't see any aggregation/composition here.

File: services/src/kilda-core/kilda-model/src/main/java/org/openkilda/utils/ICookieEnumField.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.utils;
+
+public interface ICookieEnumField {
```
Comment (sergii-iakovenko) [inline services/src/kilda-core/kilda-model/src/main/java/org/openkilda/utils/ICookieEnumField.java]: I see 2 issues here:
- Not obvious the purpose of the interface along with lack of javadoc.
- "I" prefix is against the code convention.

File: services/src/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/service/ping/packet/EthernetHeader.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    @Override
+    public byte[] serialize() {
+        byte[] payloadData = new byte[0];
+        if (payload != null) {
+            payload.setParent(this);
+            payloadData = payload.serialize();
+        }
+
+        byte[] data = new byte[payloadData.length + length];
+        ByteBuffer bb = ByteBuffer.wrap(data);
```
Comment (sergii-iakovenko) [inline services/src/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/service/ping/packet/EthernetHeader.java]: Why not use simpler ByteBuffer.allocate(payloadData.length + length) ?

File: services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/FlowPathDirection.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.model;
+
+public enum FlowPathDirection {
```
Comment (sergii-iakovenko) [inline services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/FlowPathDirection.java]: +1

File: services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/validate/OfFlowReference.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import org.openkilda.model.of.OfFlowSchema;
+
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+import java.io.Serializable;
+
+@Value
+@AllArgsConstructor
+public class OfFlowReference implements Serializable {
```
Comment (sergii-iakovenko) [inline services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/validate/OfFlowReference.java]: Where is this class used? Why do we need OF entities in the model package?

File: services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/Cookie.java [hunk 8]
```diff
@@ -257,4 +354,52 @@ public static String toString(long cookie) {
...
+                "Unable to map value %x value into %s value", needle, typeRef.getSimpleName()));
+    }
+
+    @Getter
+    static class BitField {
+        private final int width;
+        private final int offset;
+
+        public BitField(int width, int offset) {
+            if (64 < width + offset || width < 0 || offset < 0) {
```
Comment (niksv) [inline services/src/kilda-core/kilda-model/src/main/java/org/openkilda/model/Cookie.java]: Is it ok if `width  == 0`?

File: services/src/floodlight-service/floodlight-modules/src/test/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactoryTest.java [hunk 9]
```diff
@@ -92,6 +94,37 @@ public void tearDown() throws Exception {
...
+                        .setExact(MatchField.IN_PORT, OFPort.of(endpoint.getPortNumber()))
+                        .build())
+                .setInstructions(ImmutableList.of(
+                        of.instructions().applyActions(Collections.singletonList(of.actions().popVlan())),
+                        of.instructions().writeMetadata(metadata.getValue(), metadata.getMask()),
+                        of.instructions().gotoTable(TableId.of(SwitchManager.INGRESS_TABLE_ID))))
+                .build();
+        verifyOfMessageEquals(expected, factory.makeOuterVlanMatchAndRemoveMessage());
+    }
+    */
```
Comment (niksv) [inline services/src/floodlight-service/floodlight-modules/src/test/java/org/openkilda/floodlight/command/flow/ingress/of/IngressFlowModFactoryTest.java]: why it was commented? not deleted/fixed?

File: services/src/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/utils/MetadataAdapter.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import org.projectfloodlight.openflow.types.U64;
+
+public final class MetadataAdapter {
+    public static final MetadataAdapter INSTANCE = new MetadataAdapter();
+
+    // bits layout
+    // 0x01000000_00000fff - outer vlan presence flag and value
+
+    private static final U64 OUTER_VLAN_PRESENCE_FLAG = U64.of(0x01000000_00000000L);
+    private static final U64 OUTER_VLAN_MASK          = U64.of(0x00000000_00000fffL);
```
Comment (niksv) [inline services/src/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/utils/MetadataAdapter.java]: This metadata intersects with LLDP metadata from `org.openkilda.model.Metadata.java`. We need make one class with metadata (doesn't matter which) instead of two. 

PS. it will be cool it we will create some metadata scheme with types, subtypes, etc (like in Cookie.java). But we can do it in separate PR.

File: docker/floodlight-modules/src/main/java/org/openkilda/floodlight/utils/MetadataAdapter.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import org.projectfloodlight.openflow.types.U64;
+
+public final class MetadataAdapter {
+    public static final MetadataAdapter INSTANCE = new MetadataAdapter();
+
+    // bits layout
+    // 0x01000000_00000fff - outer vlan presence flag and value
+
+    private static final U64 OUTER_VLAN_PRESENCE_FLAG = U64.of(0x01000000_00000000L);
+    private static final U64 OUTER_VLAN_MASK          = U64.of(0x00000000_00000fffL);
```
Comment (niksv) [inline docker/floodlight-modules/src/main/java/org/openkilda/floodlight/utils/MetadataAdapter.java]: Noviflow switches has limitation. size of metadata is only 32 bits. (Yes in OF 1.3 spec it is 64 bits). 
So you can't use `OUTER_VLAN_PRESENCE_FLAG` flag

**Comment by surabujin** [comment]
This PR was split into set of smaller PRs. Most of them already merged. The main part was reinjected as #3438
